clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -729,7 +729,12 @@ getKernelInvocationKind(FunctionDecl *KernelCallerFunc) {
 }
 
 static const CXXRecordDecl *getKernelObjectType(FunctionDecl *Caller) {
-  return (*Caller->param_begin())->getType()->getAsCXXRecordDecl();
+  QualType KernelParamTy = (*Caller->param_begin())->getType();
```

Comment:
@elizabethandrews and @premanandrao : Did you opt to defer the diagnosing of using the wrong kernel-signature to another patch?  I don't see that comment of mine dealt with.

Reply:
@erichkeane, sorry, I mentally opted for that, but did not communicate that here.
@cperkinsintel, can you do a separate patch for that?  Pretty much every test added could potentially use the old signature, so the sooner we can get a diagnostic, the better.

Reply:
Sure, I'd be happy to - easily done.  What sort of thing are we emitting?   A  Warning I presume? What should it say? "Older SYCL headers encountered" ? 

`def warn_older_sycl_version : Warning<"Older SYCL headers encountered">`

Reply:
I think there are three warnings here.  

in 1.2.1 mode:
- Use of const & is a 'newer' sycl extension
-Use of pass-by-value is deprecated for 'newer' mode (akin to using noexcept and having it change type in C++17)
In Newer mode:
- Use of pass-by-value is deprecated as of SYCL-newer.

Reply:
How would an end-user encounter and interpret those?   We compile correctly for both old and new sycl runtime.  And the standard Clang diagnostics will warn them if they make a const violation with a functor.  

The only people writing proxy kernels is our team, for our tests. Maybe I'm misunderstanding you, but to me it seems like what we might encounter is end-user projects using SYCL 1.2 runtime headers and we should warn them of that, to encourage them to use the latest.

Reply:
The same way they interpret features based on C++ language version.  They end up being different severities of warnings in general (and of course, warnings in system headers are suppressed).

I encourage you to play around with the deprecated and extension warnings based on language features to see what I mean.

Reply:
I think we are talking past each other.  I'm not sure whether you are understanding me, or me not grokking you, or both.

For our users, people downloading DPCPP and writing SYCL code, there is no change to their code.  In SYCL 1.2:
`parallel_for<>([=](id){ .. do_something();});`
In SYCL 2020:
`parallel_for<>([=](id){ .. do_something();});`

But our runtime has changed such that we are now passing their lambda by reference, instead of by value.  And we support both in our code generation. So it is an opportunity for us to detect the older runtime and tell the user that it's not the latest. 

You seem to be approaching this from a different angle, but I'm unsure who the audience is.  Do we have a flag where the user gets to elect the SYCL version they want? (much like the `-std=c++17` flag).   If we did, then we would know what "mode" we are in and could advise them much like you describe. But, to my knowledge, we don't have a flag for them to elect anything like that.  Perhaps I just need a clearer example. 

Users writing functors, instead of lambdas, _will_ have to make a code change, because that operator needs to be `const`. But Clang will tell them this by virtue of the definition in sycl.hpp.

Reply:
I don't see the suggesting users to update their runtime as a valid warning here.  

However, my concern is that we expose sycl_kernel to non-runtime authors.  Therefore, we're going to get people writing sycl_kernel functions for themselves.

Therefore, we should warn if they're using the 'old' one.  If we DO have a flag based on the SYCL version (I think strict-sycl is our 1.2.1 compatibility flag), we should also warn in the other direction.

Reply:
I see. Hmm - as you said there is a `no-sycl-strict` flag, but it's not propagated onto the LangOpts definition, that I can see.  I'll investigate.

Reply:
Ok - I've updated the diagnostic to warn about deprecated pass-by-value.  Should be good. Let me know.

I looked into the "future feature" diagnostic as well, but that introduces two issues.  All the tests that I tried to bring up-to-date then have to be further updated with the expectation of that warning. That means I'll touch even more code.  But worse, when we finally switch to 2020 as the default version, all of those would have to be updated again.  (Or if not updated now, they will then). So that's creating a lot of busywork for ourselves.

Reply:
Typically you'd compile with the -Wno<warning> flag in those cases.  Just like the -Wc++17-extensions warning set.

Reply:
yeah, but the tests don't use that, and not sure what other consequences would be to adding it.

I think the best is to just defer the work of that "future" diagnostic until after we switch to 2020 as the default SYCL.

Reply:
As I said, I disagree.  We already HAVE switched to 2020 as the default (at least, as much as we've implemented).  As you can see, we have plenty of 2020 features by default in our compiler.

Reply:
My apologies if I wasn't specific enough. I'm talking about the context of these flags (`sycl-std`, `no-sycl-strict`), which I didn't even know existed until you told me.  

Presently using  `sycl-std=2020` with our clang is not supported.  However, in the update to this PR earlier today, I added that as a valid option.  So that's good.
But what I meant by "default": if no `sycl-std` is provided, then we elect  SYCL 1.2 (aka SYCL 2017) by default.  You can see that here: https://github.com/intel/llvm/blob/aaa575dd5b59c966ad769dfc9738c3d58a8d09f7/clang/lib/Driver/ToolChains/Clang.cpp#L4160

On a lark, I tried changing that to `sycl-std=2020` , but it causes a couple of unrelated  tests to fail. Probably for nothing serious. But me changing the default choice to of the Clang Front End 2020 in this PR which is ostensibly about changing the kernel lambda requirements in the runtime to be const references, seems like we are getting further and further afield.

I'm not the code owner, but to me switching the default to be 2020 and fixing the tests that fail as a result of that should be its own PR. And after that, I can come back and add the 'future' diagnostic you want, which, I believe, should be able to drop in with no further updates to kernel proxies in any tests since they've all been brought up to date.  Only the test that exists specifically to assert both kernel call conventions compile and emit the correct warnings will have to be updated. 

I'd be happy to go and open that other PR, for adding 2020 as an option, making it the default, etc.  I just assumed that decision was going to be coordinated with outside parties/developments. Let me know.

Reply:
Noone is asking you to change the default.  I simply want you to have a warning depending on which version we are in.  If the user selects 2020, but doesn't use references, thats a warning.  If the user uses 1.2.1 (or whatever our default is), and has a reference, thats also a warning.

Reply:
Yes, but my point is that it is easier to add the second warning after we switch the default to 2020. I'm not opposed to the warning. 
Right now, the default is 1.2.1 and most tests do not have the no-sycl-strict flag ( * ) , so if that warning is added, then a whole bunch of tests will start emitting it and need to have an expectation added, or that flag.   And when we switch the default, those tests would have to be updated again, to have that expectation removed. Alternately, I can undo the change I made to the tests bringing them up to the 2020 way of doing things. But, again, when we switch, they'll all have to be updated again.   No matter how we approach it, that second future warning causes us a lot of busywork. BUT, it's trivial to add that future warning. 

The simplest is to defer the second warning for now, and come back to it after 2020 is the default.  That's all I'm suggesting.

( * ) The clang++ driver passes the no-sycl-strict flag to the compiler. But most of these tests skip the driver and go straight to the clang_cc1  .

Reply:
Again, we use the sycl-strict flag to mean "give me warnings when I diverge from the 1.2.1 spec".  It should be a warning under sycl-strict. Which is what I said before you diverged on this.

Reply:
I am trying to implement them as you have asked.  My commit yesterday adds the 'deprecated' warning. Is this not correct? If not, let me know and I'll fix it. 
https://github.com/intel/llvm/pull/1785/commits/ed43c1bfc38eda9e9ddc99bd46b5f5f7c7b92507

I am not advocating for diverging from your request, only deferring the other warning that you are asking, because it's going to make a lot of busy work for someone else once the default changes to 2020. Let's defer the other warning you want until after 2020 is made the default for sycl-std.  It'll be less effort for everyone.

Reply:
Here's the problem: passing kernels by reference in SYCL 1.2.1 is INVALID and a strict compiler would emit this as an error.  Because of that, we need to warn, at least under sycl-strict.  I don't think we can defer that, since it is an ill-formed-allowed-as-extension diagnostic.

I think you're overstating the 'busy work'.  It is as simple as removing a bunch of -Wno-<warning> flags, though this even ends up not being a necessary change (since it doesn't actually break the tests to have a -Wno-<warning> flag that doesn't warn).

Reply:
just to be clear, when you say -Wno-  you mean the `-Wno-sycl-strict` flag, correct?

Reply:
No, you can put a warning in a warning-group.  So there would be a '-Wsycl-121-compat' group.  That is a part of the -Wsycl-strict group.

Reply:
Ahh - I didn't realize a warning group can be put into another.  Do we have another that is nested already, for me to see how its done?

Reply:
See DiagnosticGroups.td.  Theres a ton of examples.  Basically, the 'top level' group takes the lower groups.  "enum-conversion" would be the "sycl-strict" for example.

Reply:
thanks.  And that explains why I thought the flags were a big deal, and you not so much.